File: day5/ocaml/day5_1.ml
let filename = "input.txt";;

let read_lines file =
  let contents = In_channel.with_open_bin file In_channel.input_all 
  in
  String.split_on_char '\n' contents;;

type state =
  | Start
  | Seeds
  | Map;;

let process_input i: int =
  let rec helper currState strlist input (maps: (int * int * int) list list) tmpmap =
    match currState, strlist with
    | Start, x::y::xs ->  if String.starts_with ~prefix:"seeds: " x 
                          then
                            let seedlist = List.map int_of_string (String.split_on_char ' ' (String.sub x 7 (String.length x - 7)))
                            in helper Seeds xs seedlist maps tmpmap
                          else
                            failwith ("invalid input " ^ x)
    | Seeds, x::xs ->     if x = "seed-to-soil map:"
                          then helper Map xs input maps tmpmap
                          else failwith ("invalid input " ^ x)
    | Map, x::xs -> (match x with
                          | "" -> helper Map xs input (maps@[tmpmap]) []
                          | el when String.ends_with ~suffix:"map:" el -> helper Map xs input maps tmpmap
                          | el -> let maplist = List.map int_of_string (String.split_on_char ' ' el)
                                  in (match maplist with
                                    | x::y::z::[] -> helper Map xs input maps ((x, y, z)::tmpmap) 
                                    | _ -> failwith ("invalid input " ^ el))
                    )
    | Map, [] -> input, maps
    | _, _ -> failwith ("invalid input")
  and minval lis = match lis with
  | [] -> failwith "empty input"
  | x::[] -> x
  | x::xs -> let res = minval xs in if res > x then x else res
  and convertFun maplist_list el =
    let rec convertInner maplist n = 
      match maplist with
      | (dest, src, range)::xs when (src <= n) && (n < src + range) -> dest + (n - src)
      | (dest, src, range)::xs -> convertInner xs n
      | [] -> n
    in
      match maplist_list with
      | [] -> el
      | x::xs -> convertFun xs (convertInner x el) 
  in
  let i, m = helper Start i [] [] []
  in minval (List.map (convertFun m) i)

let solution = process_input(read_lines filename);;
print_endline ("Solution is " ^ (string_of_int solution));;

File: day5/python/day5_1.py
import re


with open("input.txt", "r") as file:
    lines = file.readlines()

seeds = None
current_map = list()

def process(el, cmap):
    for (dest, src, rg) in cmap:
        if src <= el and el < src + rg:
            el = dest + el - src
            break
    return el

for line in lines:
    line = line.strip()
    if line.startswith("seeds: "):
        seeds = list(map(int, line.split()[1:]))
    elif re.match("(.*)-to-(.*) map:", line) is not None:
        seeds = list(map(lambda el: process(el, current_map), seeds))
        current_map = list()
    else:
        row = re.match("([0-9]+) ([0-9]+) ([0-9]+)", line)
        if row is None:
            continue
        dest, src, rg = map(int, row.groups())
        current_map.append((dest, src, rg))

seeds = list(map(lambda el: process(el, current_map), seeds))

print(min(seeds))